go/arrow/scalar/nested.go

```diff
@@ -0,0 +1,308 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scalar
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/apache/arrow/go/arrow"
+	"github.com/apache/arrow/go/arrow/array"
+	"github.com/apache/arrow/go/arrow/internal/debug"
+	"github.com/apache/arrow/go/arrow/memory"
+	"golang.org/x/xerrors"
+)
+
+type ListScalar interface {
+	Scalar
+	List() array.Interface
+}
+
+type List struct {
+	scalar
+	Value array.Interface
+}
+
+func (l *List) value() interface{}    { return l.Value }
+func (l *List) List() array.Interface { return l.Value }
+func (l *List) equals(rhs Scalar) bool {
+	return array.ArrayEqual(l.Value, rhs.(ListScalar).List())
+}
+func (l *List) Validate() (err error) {
+	if err = l.scalar.Validate(); err != nil {
+		return
+	}
+	if err = validateOptional(&l.scalar, l.Value, "value"); err != nil {
+		return
+	}
+
+	if !l.Valid {
+		return
```

Comment:
do you want to set an err in this case?

Reply:
that case isn't an error, it just means it's a null scalar